### PR TITLE
Make Kestrel's response buffer limit configurable

### PR DESCRIPTION
**Pull request by halter73**
#1061

### CODE REVIEW

**Comment by halter73** [comment]
@cesarbs @davidfowl

File: src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs [hunk 1]
```diff
@@ -21,6 +24,30 @@ public class KestrelServerLimits
...
 
         /// <summary>
+        /// Gets or sets the maximum size of the response buffer before calls
+        /// write calls begin to block or return incomplete tasks.
+        /// </summary>
+        /// <remarks>
+        /// When set to null, the size of the response buffer is unlimited.
+        /// Defaults to 65536 bytes (64 KB).
+        /// </remarks>
+        public long? MaxResponseBufferSize
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs]: nullable huh? With great power comes great responsibility. Netty and node both have a high and low watermark used for backpressure. Outgoing writes are halted when it hits the high watermark and continue once the low watermark is hit.

Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs]: It's consistent with `MaxRequestBufferSize`.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketOutput.cs [hunk 2]
```diff
@@ -75,6 +75,7 @@ public class SocketOutput : ISocketOutput
             _tasksPending = new Queue<WaitingTask>(_initialTaskQueues);
             _writeContextPool = new Queue<WriteContext>(_maxPooledWriteContexts);
             _writeReqPool = thread.WriteReqPool;
+            _maxBytesPreCompleted = connection.ServerOptions.Limits.MaxResponseBufferSize;
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketOutput.cs]: What if it's null?

File: src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs [hunk 3]
```diff
@@ -21,6 +24,30 @@ public class KestrelServerLimits
         private int _maxRequestHeaderCount = 100;
 
         /// <summary>
+        /// Gets or sets the maximum size of the response buffer before calls
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs]: there's an extra "calls" here

File: src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs (hunk 1 above)
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs]: Sure, that has the same problem as this. When if you end up in a situation where the consumer is reading slowly, you'll end up pausing and resuming constantly. Maybe we can add it later...

File: test/Microsoft.AspNetCore.Server.KestrelTests/SocketOutputTests.cs [hunk 4]
```diff
@@ -32,9 +68,9 @@ public class SocketOutputTests
                 var socket = new MockSocket(mockLibuv, kestrelThread.Loop.ThreadId, new TestKestrelTrace());
                 var trace = new KestrelTrace(new TestKestrelTrace());
                 var ltp = new SynchronousThreadPool();
-                var socketOutput = new SocketOutput(kestrelThread, socket, new MockConnection(), "0", trace, ltp);
+                var socketOutput = new SocketOutput(kestrelThread, socket, new MockConnection(options), "0", trace, ltp);
 
-                // I doubt _maxBytesPreCompleted will ever be over a MB. If it is, we should change this test.
+                // At least one run of this test should have a MaxResponseBufferSize < 1048576.
```
Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.KestrelTests/SocketOutputTests.cs]: nit: 1 MB or 1,048,576 for easier reading

File: test/Microsoft.AspNetCore.Server.KestrelTests/SocketOutputTests.cs [hunk 5]
```diff
@@ -52,10 +88,120 @@ public class SocketOutputTests
...
+
+                var kestrelThread = kestrelEngine.Threads[0];
+                var socket = new MockSocket(mockLibuv, kestrelThread.Loop.ThreadId, new TestKestrelTrace());
+                var trace = new KestrelTrace(new TestKestrelTrace());
+                var ltp = new SynchronousThreadPool();
+                var options = new KestrelServerOptions { Limits = { MaxResponseBufferSize = null } };
+                var socketOutput = new SocketOutput(kestrelThread, socket, new MockConnection(options), "0", trace, ltp);
+
+                // Don't want to allocate anything too huge for perf. This is at least larger than the default buffer.
+                var bufferSize = 1048576;
```
Comment (cesarbs) [inline test/Microsoft.AspNetCore.Server.KestrelTests/SocketOutputTests.cs]: nit: `1024 * 1024`, again for easier reading

File: src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs [hunk 6]
```diff
@@ -21,6 +24,30 @@ public class KestrelServerLimits
...
+        /// write calls begin to block or return incomplete tasks.
+        /// </summary>
+        /// <remarks>
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/KestrelServerLimits.cs]: Add remark about 0 disabling response buffering.

File: src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketOutput.cs [hunk 7]
```diff
@@ -51,7 +51,7 @@ public class SocketOutput : ISocketOutput
         private bool _postingWrite = false;
 
         private bool _cancelled = false;
-        private int _numBytesPreCompleted = 0;
+        private long _numBytesPreCompleted = 0;
```
Comment (cesarbs) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketOutput.cs]: One thing to watch out for here is that `long` is not guaranteed to have atomic reads and writes. Do we ever read and write this concurrently?

Comment (davidfowl) [inline src/Microsoft.AspNetCore.Server.Kestrel/Internal/Http/SocketOutput.cs]: It's updated under the `_contextLock`

**Pull request merged**